### ISSUE TITLE
Twig taint analysis fails when PHP file referencing twig file is analyzed first

### ISSUE THREAD

**Issue by ostrolucky** [psalm/psalm-plugin-symfony#84]
Plugin keeps crashing on

```
return $this->container->get('twig')->render('AppBundle:DataProvider/GraduateJobs:job.xml.twig', [...]);
```

with error

```
Uncaught RuntimeException: The template AppBundle:DataProvider/GraduateJobs:job.xml.twig was not found. in /var/www/vendor/psalm/plugin-symfony/src/Twig/CachedTemplatesMapping.php:67
Stack trace:
#0 /var/www/vendor/psalm/plugin-symfony/src/Twig/CachedTemplatesTainter.php(58): Psalm\SymfonyPsalmPlugin\Twig\CachedTemplatesMapping::getCacheClassName('AppBundle:DataP...')
#1 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Provider/MethodReturnTypeProvider.php(105): Psalm\SymfonyPsalmPlugin\Twig\CachedTemplatesTainter::getMethodReturnType(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), 'Twig\\Environmen...', 'render', Array, Object(Psalm\Context), Object(Psalm\CodeLocation), NULL, NULL, NULL)
#2 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/Call/Method/MethodCallReturnTypeFetcher.php(53): Psalm\Internal\Provider\MethodReturnTypeProvider->getReturnType(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), 'Twig\\Environmen...', 'render', Array, Object(Psalm\Context), Object(Psalm\CodeLocation), NULL)
#3 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/Call/Method/AtomicMethodCallAnalyzer.php(680): Psalm\Internal\Analyzer\Statements\Expression\Call\Method\MethodCallReturnTypeFetcher::fetch(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(Psalm\Codebase), Object(PhpParser\Node\Expr\MethodCall), Object(Psalm\Context), Object(Psalm\Internal\MethodIdentifier), Object(Psalm\Internal\MethodIdentifier), Object(Psalm\Internal\MethodIdentifier), 'Twig\\Environmen...', Object(Psalm\Type\Atomic\TNamedObject), Object(Psalm\Type\Atomic\TNamedObject), Array, Object(Psalm\Internal\Analyzer\Statements\Expression\Call\Method\AtomicMethodCallAnalysisResult), Object(Psalm\Internal\Type\TemplateResult))
#4 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/Call/MethodCallAnalyzer.php(168): Psalm\Internal\Analyzer\Statements\Expression\Call\Method\AtomicMethodCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\MethodCall), Object(Psalm\Codebase), Object(Psalm\Context), Object(Psalm\Type\Atomic\TNamedObject), Object(Psalm\Type\Atomic\TNamedObject), false, NULL, Object(Psalm\Internal\Analyzer\Statements\Expression\Call\Method\AtomicMethodCallAnalysisResult))
#5 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php(144): Psalm\Internal\Analyzer\Statements\Expression\Call\MethodCallAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\MethodCall), Object(Psalm\Context))
#6 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php(39): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::handleExpression(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\MethodCall), Object(Psalm\Context), false, NULL, false)
#7 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/ReturnAnalyzer.php(135): Psalm\Internal\Analyzer\Statements\ExpressionAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Expr\MethodCall), Object(Psalm\Context))
#8 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(457): Psalm\Internal\Analyzer\Statements\ReturnAnalyzer::analyze(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Return_), Object(Psalm\Context))
#9 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php(164): Psalm\Internal\Analyzer\StatementsAnalyzer::analyzeStatement(Object(Psalm\Internal\Analyzer\StatementsAnalyzer), Object(PhpParser\Node\Stmt\Return_), Object(Psalm\Context), Object(Psalm\Context))
#10 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FunctionLikeAnalyzer.php(591): Psalm\Internal\Analyzer\StatementsAnalyzer->analyze(Array, Object(Psalm\Context), Object(Psalm\Context), true)
#11 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(1897): Psalm\Internal\Analyzer\FunctionLikeAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Internal\Provider\NodeDataProvider), Object(Psalm\Context))
#12 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(734): Psalm\Internal\Analyzer\ClassAnalyzer->analyzeClassMethod(Object(PhpParser\Node\Stmt\ClassMethod), Object(Psalm\Storage\ClassLikeStorage), Object(Psalm\Internal\Analyzer\ClassAnalyzer), Object(Psalm\Context), Object(Psalm\Context))
#13 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(215): Psalm\Internal\Analyzer\ClassAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Context))
#14 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(336): Psalm\Internal\Analyzer\FileAnalyzer->analyze(NULL)
#15 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(576): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(227, '/var/www/src/Ap...')
#16 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(268): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1)
#17 /var/www/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(639): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 1, false, true)
#18 /var/www/vendor/vimeo/psalm/src/psalm.php(677): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/var/www/', false)
#19 /var/www/vendor/vimeo/psalm/psalm(2): require_once('/var/www/vendor...')
#20 {main}
(Psalm 3.16@d03e5ef057d6adc656c0ff7e166c50b73b4f48f3 crashed due to an uncaught Throwable)
```

even though there is such cache file present in `var/cache/dev/twig/6c/6c3e07fca810aa921fb8bc653998ccb1b777f9ed15f04263962af77578011498.php`: 
```php
<?php

use Twig\Environment;
use Twig\Error\LoaderError;
use Twig\Error\RuntimeError;
use Twig\Extension\SandboxExtension;
use Twig\Markup;
use Twig\Sandbox\SecurityError;
use Twig\Sandbox\SecurityNotAllowedTagError;
use Twig\Sandbox\SecurityNotAllowedFilterError;
use Twig\Sandbox\SecurityNotAllowedFunctionError;
use Twig\Source;
use Twig\Template;

/* AppBundle:DataProvider/GraduateJobs:job.xml.twig */
class __TwigTemplate_943d73236d880c64fb3bff57201b2a5745c3c2ea7949e6cdfc27cd8cd2a37f48 extends \Twig\Template
{
    private $source;
    private $macros = [];

    public function __construct(Environment $env)
    {
        parent::__construct($env);

        $this->source = $this->getSourceContext();

        $this->parent = false;

        $this->blocks = [
        ];
    }

    protected function doDisplay(array $context, array $blocks = [])
    {
        $macros = $this->macros;
        $__internal_085b0142806202599c7fe3b329164a92397d8978207a37e79d70b8c52599e33e = $this->extensions["Symfony\\Bundle\\WebProfilerBundle\\Twig\\WebProfilerExtension"];
        $__internal_085b0142806202599c7fe3b329164a92397d8978207a37e79d70b8c52599e33e->enter($__internal_085b0142806202599c7fe3b329164a92397d8978207a37e79d70b8c52599e33e_prof = new \Twig\Profiler\Profile($this->getTemplateName(), "template", "AppBundle:DataProvider/GraduateJobs:job.xml.twig"));

        $__internal_319393461309892924ff6e74d6d6e64287df64b63545b994e100d4ab223aed02 = $this->extensions["Symfony\\Bridge\\Twig\\Extension\\ProfilerExtension"];
        $__internal_319393461309892924ff6e74d6d6e64287df64b63545b994e100d4ab223aed02->enter($__internal_319393461309892924ff6e74d6d6e64287df64b63545b994e100d4ab223aed02_prof = new \Twig\Profiler\Profile($this->getTemplateName(), "template", "AppBundle:DataProvider/GraduateJobs:job.xml.twig"));

        // line 1
        echo "<?xml version = '1.0' encoding = 'utf-8' ?>
<!DOCTYPE job PUBLIC 'gj-job-dtd' 'http://www.graduate-jobs.com/xml/job.dtd'>
<job>
    <status>";
        // line 4
        echo twig_escape_filter($this->env, (isset($context["workMode"]) || array_key_exists("workMode", $context) ? $context["workMode"] : (function () { throw new RuntimeError('Variable "workMode" does not exist.', 4, $this->source); })()), "html", null, true);
        echo "</status>
    <accountName>";
        // line 5
        echo twig_escape_filter($this->env, (isset($context["accountName"]) || array_key_exists("accountName", $context) ? $context["accountName"] : (function () { throw new RuntimeError('Variable "accountName" does not exist.', 5, $this->source); })()), "html", null, true);
        echo "</accountName>
    <password>";
        // line 6
        echo twig_escape_filter($this->env, (isset($context["password"]) || array_key_exists("password", $context) ? $context["password"] : (function () { throw new RuntimeError('Variable "password" does not exist.', 6, $this->source); })()), "html", null, true);
        echo "</password>
    <title>";
        // line 7
        echo twig_escape_filter($this->env, (isset($context["title"]) || array_key_exists("title", $context) ? $context["title"] : (function () { throw new RuntimeError('Variable "title" does not exist.', 7, $this->source); })()), "html", null, true);
        echo "</title>
    <yourRef>";
        // line 8
        echo twig_escape_filter($this->env, (isset($context["yourRef"]) || array_key_exists("yourRef", $context) ? $context["yourRef"] : (function () { throw new RuntimeError('Variable "yourRef" does not exist.', 8, $this->source); })()), "html", null, true);
        echo "</yourRef>
    <keyword>";
        // line 9
        echo twig_escape_filter($this->env, (isset($context["keyword"]) || array_key_exists("keyword", $context) ? $context["keyword"] : (function () { throw new RuntimeError('Variable "keyword" does not exist.', 9, $this->source); })()), "html", null, true);
        echo "</keyword>
    <detail>";
        // line 10
        echo twig_escape_filter($this->env, (isset($context["detail"]) || array_key_exists("detail", $context) ? $context["detail"] : (function () { throw new RuntimeError('Variable "detail" does not exist.', 10, $this->source); })()), "html", null, true);
        echo "</detail>
    ";
        // line 11
        $context['_parent'] = $context;
        $context['_seq'] = twig_ensure_traversable((isset($context["locations"]) || array_key_exists("locations", $context) ? $context["locations"] : (function () { throw new RuntimeError('Variable "locations" does not exist.', 11, $this->source); })()));
        foreach ($context['_seq'] as $context["_key"] => $context["location"]) {
            // line 12
            echo "        <location>";
            echo twig_escape_filter($this->env, $context["location"], "html", null, true);
            echo "</location>
    ";
        }
        $_parent = $context['_parent'];
        unset($context['_seq'], $context['_iterated'], $context['_key'], $context['location'], $context['_parent'], $context['loop']);
        $context = array_intersect_key($context, $_parent) + $_parent;
        // line 14
        echo "    ";
        if (((isset($context["postcodes"]) || array_key_exists("postcodes", $context) ? $context["postcodes"] : (function () { throw new RuntimeError('Variable "postcodes" does not exist.', 14, $this->source); })()) != "")) {
            // line 15
            echo "        <postcodes>";
            echo twig_escape_filter($this->env, (isset($context["postcodes"]) || array_key_exists("postcodes", $context) ? $context["postcodes"] : (function () { throw new RuntimeError('Variable "postcodes" does not exist.', 15, $this->source); })()), "html", null, true);
            echo "</postcodes>
    ";
        }
        // line 17
        echo "    <salary>";
        echo twig_escape_filter($this->env, (isset($context["salary"]) || array_key_exists("salary", $context) ? $context["salary"] : (function () { throw new RuntimeError('Variable "salary" does not exist.', 17, $this->source); })()), "html", null, true);
        echo "</salary>
    <displayEndDate>";
        // line 18
        echo twig_escape_filter($this->env, (isset($context["displayEndDate"]) || array_key_exists("displayEndDate", $context) ? $context["displayEndDate"] : (function () { throw new RuntimeError('Variable "displayEndDate" does not exist.', 18, $this->source); })()), "html", null, true);
        echo "</displayEndDate>
    ";
        // line 19
        $context['_parent'] = $context;
        $context['_seq'] = twig_ensure_traversable((isset($context["sectors"]) || array_key_exists("sectors", $context) ? $context["sectors"] : (function () { throw new RuntimeError('Variable "sectors" does not exist.', 19, $this->source); })()));
        foreach ($context['_seq'] as $context["_key"] => $context["sector"]) {
            // line 20
            echo "        <industrySector>";
            echo twig_escape_filter($this->env, $context["sector"], "html", null, true);
            echo "</industrySector>
    ";
        }
        $_parent = $context['_parent'];
        unset($context['_seq'], $context['_iterated'], $context['_key'], $context['sector'], $context['_parent'], $context['loop']);
        $context = array_intersect_key($context, $_parent) + $_parent;
        // line 22
        echo "    <applyTo>";
        echo twig_escape_filter($this->env, (isset($context["applyTo"]) || array_key_exists("applyTo", $context) ? $context["applyTo"] : (function () { throw new RuntimeError('Variable "applyTo" does not exist.', 22, $this->source); })()), "html", null, true);
        echo "</applyTo>
    ";
        // line 23
        $context['_parent'] = $context;
        $context['_seq'] = twig_ensure_traversable((isset($context["sites"]) || array_key_exists("sites", $context) ? $context["sites"] : (function () { throw new RuntimeError('Variable "sites" does not exist.', 23, $this->source); })()));
        foreach ($context['_seq'] as $context["_key"] => $context["site"]) {
            // line 24
            echo "        <site>";
            echo twig_escape_filter($this->env, $context["site"], "html", null, true);
            echo "</site>
    ";
        }
        $_parent = $context['_parent'];
        unset($context['_seq'], $context['_iterated'], $context['_key'], $context['site'], $context['_parent'], $context['loop']);
        $context = array_intersect_key($context, $_parent) + $_parent;
        // line 26
        echo "    <gradeRequired>";
        echo twig_escape_filter($this->env, (isset($context["gradeRequired"]) || array_key_exists("gradeRequired", $context) ? $context["gradeRequired"] : (function () { throw new RuntimeError('Variable "gradeRequired" does not exist.', 26, $this->source); })()), "html", null, true);
        echo "</gradeRequired>
</job>
";
        
        $__internal_085b0142806202599c7fe3b329164a92397d8978207a37e79d70b8c52599e33e->leave($__internal_085b0142806202599c7fe3b329164a92397d8978207a37e79d70b8c52599e33e_prof);

        
        $__internal_319393461309892924ff6e74d6d6e64287df64b63545b994e100d4ab223aed02->leave($__internal_319393461309892924ff6e74d6d6e64287df64b63545b994e100d4ab223aed02_prof);

    }

    public function getTemplateName()
    {
        return "AppBundle:DataProvider/GraduateJobs:job.xml.twig";
    }

    public function isTraitable()
    {
        return false;
    }

    public function getDebugInfo()
    {
        return array (  138 => 26,  129 => 24,  125 => 23,  120 => 22,  111 => 20,  107 => 19,  103 => 18,  98 => 17,  92 => 15,  89 => 14,  80 => 12,  76 => 11,  72 => 10,  68 => 9,  64 => 8,  60 => 7,  56 => 6,  52 => 5,  48 => 4,  43 => 1,);
    }

    public function getSourceContext()
    {
        return new Source("<?xml version = '1.0' encoding = 'utf-8' ?>
<!DOCTYPE job PUBLIC 'gj-job-dtd' 'http://www.graduate-jobs.com/xml/job.dtd'>
<job>
    <status>{{ workMode }}</status>
    <accountName>{{ accountName }}</accountName>
    <password>{{ password }}</password>
    <title>{{ title }}</title>
    <yourRef>{{ yourRef }}</yourRef>
    <keyword>{{ keyword }}</keyword>
    <detail>{{ detail }}</detail>
    {% for location in locations %}
        <location>{{ location }}</location>
    {% endfor %}
    {%  if postcodes != '' %}
        <postcodes>{{ postcodes }}</postcodes>
    {% endif %}
    <salary>{{ salary }}</salary>
    <displayEndDate>{{ displayEndDate }}</displayEndDate>
    {% for sector in sectors %}
        <industrySector>{{ sector }}</industrySector>
    {% endfor %}
    <applyTo>{{ applyTo }}</applyTo>
    {% for site in sites %}
        <site>{{ site }}</site>
    {% endfor %}
    <gradeRequired>{{ gradeRequired }}</gradeRequired>
</job>
", "AppBundle:DataProvider/GraduateJobs:job.xml.twig", "/var/www/src/AppBundle/Resources/views/DataProvider/GraduateJobs/job.xml.twig");
    }
}
```

Happy to assist with debugging the issue, but right now I am out of ideas, as I don't know how is this part designed to work.

Here's relevant part of config:
```xml
     <plugins>
        <pluginClass class="Psalm\SymfonyPsalmPlugin\Plugin">
            <containerXml>var/cache/dev/srcKernelDevDebugContainer.xml</containerXml>
            <twigCachePath>var/cache/dev/twig</twigCachePath>
        </pluginClass>
        <pluginClass class="Weirdan\DoctrinePsalmPlugin\Plugin"/>
    </plugins>

    <fileExtensions>
        <extension name=".php" />
        <extension name=".twig" checker="./vendor/psalm/plugin-symfony/src/Twig/TemplateFileAnalyzer.php"/>
    </fileExtensions>
```

**Comment by seferov** [comment]
@ostrolucky thanks for reporting.

@adrienlucas do you have time to look at it?

**Comment by adrienlucas** [comment]
Hey ! I tested with a `.xml.twig` scenario and it seemed to go well.

Anyway, it should be the `CachedTemplatesMapping` not doing its job.

@ostrolucky do you have an `ignoreFiles` that is excluding the `cache` directory in the rest of your config ?

**Comment by ostrolucky** [comment]
No, I don't exclude cache directory. Here is relevant section:
```xml
<projectFiles>
        <directory name="src"/>
        <directory name="app"/>
        <ignoreFiles>
            <directory name="vendor"/>
            <directory name="src/AppBundle/Tests/_support" />
        </ignoreFiles>
    </projectFiles>
```

From my view it looks like `GraduateJobs.php`is parsed & analyzed first and only after that `job.xml.twig`. This means when analyzing php file, twig file was not analyzed yet, so it's missing in template cache. What's guaranteeing order of analyzing here, why does this plugin assume twig files are analyzed first?

**Comment by adrienlucas** [comment]
Oh you are right ! Thanks for catching it, I did not manage to reproduce it at the time in the tests scenarios and forgot about it.

I'm working on a fix !

**Comment by adrienlucas** [comment]
@ostrolucky if you could try this : https://github.com/psalm/psalm-plugin-symfony/pull/89

**Comment by ostrolucky** [comment]
Yep that solves this issue, thanks :) It uncovers another one, but I should probably create separate issue for that.
